### PR TITLE
Compat update

### PR DESCRIPTION
**Pull request by EPuncker**
lets make it friendly to 0.4 users to migrate to 1.x

[0.4 [luascript.cpp if you need to know what a certain function does](https://github.com/Fir3element/3777/blob/master/src/luascript.cpp)]

## special reviews needed:
- check if doPlayerSetSkillLevel parameters are correct
- getCreatureStorage should keep uid or change it do cid?
- check if doPlayerTransferMoneyTo is correct
- doSummonMonster(cid, name) - needs to be tested!
- check the whole Player.getWeaponType() function
- doPlayerSetPromotionLevel 'level' parameter is not used, how to handle?

### CODE REVIEW

**Comment by luanluciano93** [comment]
In the functions **getPlayerGUIDByName** and **getAccountNumberByPlayerName** replace the **getDataInt** with **getNumber**

**Comment by luanluciano93** [comment]
In data/globalevents/scripts/record.lua change **broadcastMessage** with **Game.broadcastMessage**

**Comment by EPuncker** [comment]
done

File: data/lib/compat/compat.lua [hunk 1]
```diff
@@ -567,13 +604,26 @@ getPlayerAccountBalance = getPlayerBalance
...
 function doPlayerAddMoney(cid, money) local p = Player(cid) return p and p:addMoney(money) or false end
 function doPlayerRemoveMoney(cid, money) local p = Player(cid) return p and p:removeMoney(money) or false end
+function doPlayerTransferMoneyTo(cid, target, money)
+	if not isValidMoney(money) then
+		return false
+	end
+	local p = Player(cid)
+	return p and p:transferMoneyTo(target, money) or false
+end
+function doPlayerSave(cid, ...) local p = Player(cid) return p and p:savePlayer(shallow or false) or false end
```
Comment (nekiro) [inline data/lib/compat/compat.lua:621]: player:save()?

File: data/lib/compat/compat.lua [hunk 2]
```diff
@@ -567,13 +604,26 @@ getPlayerAccountBalance = getPlayerBalance
 getIpByName = getIPByPlayerName
 
 function setPlayerStorageValue(cid, key, value) local p = Player(cid) return p and p:setStorageValue(key, value) or false end
+function doPlayerSetNameDescription(cid, desc) local p = Player(cid) return p and p:setDescription(desc) or false end
+function doPlayerSetMaxCapacity(cid, cap) local p = Player(cid) return p and p:setMaxCap(cap) or false end
```
Comment (nekiro) [inline data/lib/compat/compat.lua:608]: setMaxCap does not exist

File: data/lib/compat/compat.lua [hunk 3]
```diff
@@ -567,13 +604,26 @@ getPlayerAccountBalance = getPlayerBalance
...
+function doPlayerSetSpecialDescription(cid, desc) local p = Player(cid) return p and p:setSpecialDescription(desc) or false end
```
Comment (nekiro) [inline data/lib/compat/compat.lua:609]: setSpecialDescription does not exist

File: data/lib/core/player.lua [hunk 4]
```diff
@@ -203,3 +203,34 @@ function Player.depositMoney(self, amount)
 	self:setBankBalance(self:getBankBalance() + amount)
 	return true
 end
+
+function Player.addLevel(self, amount, round)
+	local experience, level, amount = 0, self:getLevel(), amount or 1
+	if(amount > 0) then
```
Comment (nekiro) [inline data/lib/core/player.lua:209]: if amount > 0 then

File: data/lib/core/player.lua [hunk 5]
```diff
@@ -203,3 +203,34 @@ function Player.depositMoney(self, amount)
...
+		return self:addLevel(value, round)
+	elseif skillId == SKILL_MAGLEVEL then
+		return self:addMagicLevel(value)
+	end
+	return self:addSkillTries(skillId, self:getVocation():getRequiredSkillTries(skillId, self:getSkillLevel(skillId) + value) - self:getSkillTries(skillId))
+end
+
+function Player.getWeaponType()
+	local weapon = self:getSlotItem(CONST_SLOT_LEFT)
+	if (weapon) then
```
Comment (nekiro) [inline data/lib/core/player.lua:232]: if weapon then

**Comment by luanluciano93** [comment]
Change **isItemFluidContainer(itemid)** to ### ItemType(itemId):isFluidContainer()

> https://github.com/otland/forgottenserver/blob/ec0fd327146fe51c35165c57fc564a47fbad160f/data/npc/lib/npcsystem/modules.lua#L1001

**Comment by EPuncker** [comment]
> ItemType(itemId):isFluidContainer()

thanks, keep it coming 😄

**Comment by EPuncker** [comment]
this is pretty much ready, and I wanted to push it to 1.4 since it is kinda useful in order to make it easier for people to migrate from 0.x to 1.x so I'd like to get a review from developers

**Comment by DSpeichert** [comment]
Is there any specific 0.4-style datapack you can find that wouldn't normally work with TFS master version but after this change it would? That would be a good test case and great example for when we release next version.

File: data/global.lua [hunk 6]
```diff
@@ -164,3 +164,8 @@ function getPlayerDatabaseInfo(name_or_guid)
 	result.free(query)
 	return info
 end
+
+function getExperienceForLevel(level)
+	local level = level - 1
+	return (((50 * level * level * level) - (150 * level * level) + (400 * level))/3)
```
Comment (ranisalt) [inline data/global.lua:170]: Parenthesis are inconsistent, and most of them are not needed at all.

```lua
return (50 * level * level * level - 150 * level * level + 400 * level) / 3
```

**Review by ranisalt — CHANGES_REQUESTED** [review]

**Comment by EPuncker** [comment]
> Is there any specific 0.4-style datapack you can find that wouldn't normally work with TFS master version but after this change it would? That would be a good test case and great example for when we release next version.

no, I made it specially by searching function names on otland search and seeing usage and errors

**Review by ranisalt — DISMISSED** [review]
:+1:

**Review by nekiro — COMMENTED** [review]
just trying to unlock my request changes

**Review by nekiro — APPROVED** [review]
just ublocking my review,

**Review by Znote — APPROVED** [review]

File: data/lib/compat/compat.lua [hunk 7]
```diff
@@ -1430,3 +1430,6 @@ end
 function doPlayerSendChannelMessage(cid, author, message, SpeakClasses, channel)
 	return Player(cid):sendChannelMessage(author, message, SpeakClasses, channel)
 end
+
+-- this is a fix for lua52 or higher which has the function renamed to table.unpack, while luajit still uses unpack
+if unpack == nil then unpack = table.unpack end
```
Comment (DSpeichert) [inline data/lib/compat/compat.lua:1435]: nice!

**Review by DSpeichert — APPROVED** [review]

**Pull request merged**